### PR TITLE
Wordpiecetokenizer vocabulary_size argument

### PR DESCRIPTION
**Pull request by blackhat-coder**
Closes [156](https://github.com/keras-team/keras-nlp/issues/156)

### CODE REVIEW

File: keras_nlp/tokenizers/word_piece_tokenizer.py [hunk 1]
```diff
@@ -12,17 +12,17 @@
...
 from typing import Iterable
 from typing import List
 from typing import Union
-
-import tensorflow as tf
-import tensorflow_text as tf_text
-
 from keras_nlp.tokenizers import tokenizer
 
+
```
Comment (mattdangerw) [inline keras_nlp/tokenizers/word_piece_tokenizer.py:25]: Remove

File: keras_nlp/tokenizers/word_piece_tokenizer.py [hunk 2]
```diff
@@ -12,17 +12,17 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import tensorflow as tf
```
Comment (mattdangerw) [inline keras_nlp/tokenizers/word_piece_tokenizer.py:15]: I don't think the ordering of these imports should change, and you are getting errors on CI. You may not be running the format and lint scripts correctly.

https://github.com/keras-team/keras-nlp/runs/6634310403?check_suite_focus=true

File: keras_nlp/tokenizers/word_piece_tokenizer.py [hunk 3]
```diff
@@ -199,13 +200,35 @@ def __init__(
 
         super().__init__(**kwargs)
 
+        if vocabulary_size != None and isinstance(vocabulary_size, int) == False:
```
Comment (mattdangerw) [inline keras_nlp/tokenizers/word_piece_tokenizer.py:203]: I think we can probably do without this error, we don't have errors for `lowercase` being a bool for example.

File: keras_nlp/tokenizers/word_piece_tokenizer.py [hunk 4]
```diff
@@ -199,13 +200,35 @@ def __init__(
...
+            raise ValueError(
+                f"Vocabulary_size expected type int, given {type(vocabulary_size)}"
+            )
+        
         if isinstance(vocabulary, str):
-            self.vocabulary = [
-                line.rstrip() for line in tf.io.gfile.GFile(vocabulary)
-            ]
+            if vocabulary_size is None:
```
Comment (mattdangerw) [inline keras_nlp/tokenizers/word_piece_tokenizer.py:209]: We should move all of this logic out of the file case. Setting a vocabulary_size should also truncate a vocabulary that was passed as a list. There should be no difference in the functionality (or the truncation error message) if vocab is a file or vocab is a list.

Totally fine to just do this as a post process, where we read the entire file, and do the truncation afterward.

File: keras_nlp/tokenizers/word_piece_tokenizer.py [hunk 5]
```diff
@@ -228,6 +251,7 @@ def __init__(
         self.keep_pattern = keep_pattern
         self.suffix_indicator = suffix_indicator
         self.oov_token = oov_token
+        self.vocabulary_size = vocabulary_size
```
Comment (mattdangerw) [inline keras_nlp/tokenizers/word_piece_tokenizer.py:254]: We should also add this to `get_config`

File: keras_nlp/tokenizers/word_piece_tokenizer_test.py [hunk 6]
```diff
@@ -41,6 +39,16 @@ def test_dense_output(self):
...
 
+    def test_vocabulary_size(self):
+        input_data = ["the quick brown fox."]
+        vocab_data = ["[UNK]", "the", "qu", "##ick", "br", "##own", "fox", "."]
+        vocab_size = 5
+        tokenizer = WordPieceTokenizer(
+            vocabulary=vocab_data, vocabulary_size=vocab_size
+        )
+        call_output = tokenizer(input_data)
+        self.assertEqual(tokenizer.vocabulary_size_(), vocab_size)
```
Comment (mattdangerw) [inline keras_nlp/tokenizers/word_piece_tokenizer_test.py:50]: why all the `vocabulary_size_` here. Is this a typo?

File: keras_nlp/tokenizers/word_piece_tokenizer_test.py [hunk 7]
```diff
@@ -183,6 +191,30 @@ def test_from_file(self):
...
+        vocab_path = os.path.join(self.get_temp_dir(), "vocab.txt")
+        input_data = ["the quick brown fox."]
+        vocab_data = ["[UNK]", "the", "qu", "##ick", "br", "##own", "fox", "."]
+        vocab_size = 3
+        with tf.io.gfile.GFile(vocab_path, "w") as file:
+            for piece in vocab_data:
+                file.write(piece + "\n")
+        tokenizer = WordPieceTokenizer(vocabulary=vocab_path, vocabulary_size=vocab_size)
+        call_output = tokenizer(input_data)
+        self.assertAllEqual(tokenizer.vocabulary_size_(), vocab_size)
```
Comment (mattdangerw) [inline keras_nlp/tokenizers/word_piece_tokenizer_test.py:205]: just collapse these two test into one, and assert both the layer output and the size of the vocabulary in the same test.

File: keras_nlp/tokenizers/word_piece_tokenizer_test.py [hunk 8]
```diff
@@ -41,6 +39,16 @@ def test_dense_output(self):
...
         self.assertAllEqual(call_output, [[1, 2, 3, 4, 5, 6, 7, 0, 0, 0]])
 
+    def test_vocabulary_size(self):
+        input_data = ["the quick brown fox."]
+        vocab_data = ["[UNK]", "the", "qu", "##ick", "br", "##own", "fox", "."]
+        vocab_size = 5
+        tokenizer = WordPieceTokenizer(
+            vocabulary=vocab_data, vocabulary_size=vocab_size
+        )
+        call_output = tokenizer(input_data)
```
Comment (mattdangerw) [inline keras_nlp/tokenizers/word_piece_tokenizer_test.py:49]: You should actually check this output with an assert.

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks for the PR! Left some initial comments

File: keras_nlp/tokenizers/word_piece_tokenizer_test.py (hunk 6 above)
Comment (blackhat-coder) [reply keras_nlp/tokenizers/word_piece_tokenizer_test.py:50]: The method `vocabulary_size()` has the same name with `vocabulary_size`. This gives a TypeError when  `vocabulary_size()` is called.

Comment (mattdangerw) [reply keras_nlp/tokenizers/word_piece_tokenizer_test.py:50]: Oh got it. We should preserve the `vocabulary_size()` function name, and just leave `self._vocabulary_size` as the variable name.

**Review by mattdangerw — CHANGES_REQUESTED** [review]

**Comment by mattdangerw** [comment]
@blackhat-coder any update on this?

**Pull request closed without merging**